modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java

```diff
@@ -1801,4 +1842,178 @@ private boolean enterBusy() {
+            });
+    }
+
+    /**
+     * Removes metrics for service {@code srvcName}.
+     *
+     * @param srvcName Service name.
+     */
+    private void unregisterMetrics(String srvcName) {
+        ctx.metric().remove(serviceMetricRegistryName(srvcName));
+
+        invocationHistograms.remove(srvcName);
+    }
+
+    /**
+     * Creates histogram for service method. If exist,s considers one or several argument types has same name but
+     * different package and tries to extend metric name with abbreviation of java package name.
+     *
+     * @param srvcName Service name.
+     * @param method Method to measure.
+     * @return Histogram of service method timings.
+     */
+    HistogramMetricImpl createHistogram(String srvcName, Method method) {
+        MetricRegistry metricRegistry = ctx.metric().registry(serviceMetricRegistryName(srvcName));
+
+        HistogramMetricImpl histogram = null;
+
+        // Find/create histogram.
+        for (int i = 0; i <= MAX_ABBREVIATE_NAME_LVL; ++i) {
+            String methodMetricName = methodMetricName(method, i);
+
+            synchronized (metricRegistry) {
+                // If the metric exists skip and try extending metric name in next cycle.
+                if (metricRegistry.findMetric(methodMetricName) == null) {
+                    histogram = metricRegistry.histogram(methodMetricName, DEFAULT_INVOCATION_BOUNDS,
+                        DESCRIPTION_OF_INVOCATION_METRIC);
+
+                    break;
+                }
+            }
+        }
+
+        assert histogram != null;
+
+        return histogram;
+    }
+
+    /**
+     * @param method Method for the invocation timings.
+     * @param pkgNameDepth Level of package name abbreviation. See {@link MetricUtils#abbreviateName(Class, int)}.
+     * @return Metric name for {@code method}.
+     */
+    static String methodMetricName(Method method, int pkgNameDepth) {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append(abbreviateName(method.getReturnType(), pkgNameDepth));
+        sb.append(" ");
+        sb.append(method.getName());
+        sb.append("(");
+        sb.append(Stream.of(method.getParameterTypes()).map(t -> abbreviateName(t, pkgNameDepth))
+            .collect(Collectors.joining(", ")));
+        sb.append(")");
+
+        return sb.toString();
+    }
+
+    /** @return {@code True} if metrics should not be created for this class or interface. */
+    private static boolean isMetricIgnoredFor(Class<?> cls){
+        return Object.class.equals(cls) || Service.class.equals(cls) || Externalizable.class.equals(cls);
+    }
+
+    /**
+     * Searches histogram for service method.
+     *
+     * @param srvcName Service name.
+     * @param mtd Service method.
+     * @return Histogram for {@code srvcName} and {@code mtd} or {@code null} if not found.
+     */
+    HistogramMetricImpl histogram(String srvcName, Method mtd) {
```

Comment:
@ivandasch , I agree with the code simplifying. I'm against synchronization and `ConcurrentHashMap` without certain case because metrics should be quick and do not affect invocation of target service method. Here is no iteration over the map and parallel puts. So, parallel gets from several threads are safe. Why do we need any synchronization here? 
`HistogramHolder#addIfAbsent()` is synchronized because there are several put/assigns/writes. And concurrent calls made from `IgniteServiceProcessor.undeploy()` and `IgniteServiceProcessor.redeploy()` are possible.

Reply:
As I wrote above, there is no guarantee that `srvHistos.get` would work correctly and return something meaningfull without synchronization. Synchronized block guarantees that if we obtain monitor  and read from map after releasing it after write to map, that write **happened-before** read.

Reply:
I'm talking about whole usage, not certain method. This method is not a public API, not an internal API. It is called from one place. Result of this method is not cached into any variable. What may happen:
Service method is being invoked concurrently with metrics registration. Service is already initialized. The method works. Metrics is already actually put, but `srvHistos.get` returned null. Well, we just skip measuring a coulpe of calls. It is rare case and doesn't affect whole histogram. What could be wrong here? What I missed?

Reply:
Putting in HashMap is not an atomic operation. There is no guarantee, that when performing get in another thread, this thread sees underlying array and bucket linked list in consistent state. Returning null is a one of good scenarious. Things may be even worse.

Reply:
As far as I know concurrent put/get/remove is not atomic but is safe meaning won't fail, throw an exception. You can only get null when a value exist, or a value when it is already removed.

Please, take a look to this simple test. Forks for me. Gives null periodically, and, I believe, inconsistent values. But no any lock/crash scenario. What else should we check?

```
@Test
    public void testMapConcurrency() throws InterruptedException {
        final Map<Integer, Integer> map = new HashMap<>(1);
        final Random rnd = new Random();
        final int valueCnt = 100000;
        final Integer[] holder = new Integer[1];

        for(int i=0; i<5; ++i) {
            new Thread(() -> {
                while (true) {
                    for (int v = 0; v < valueCnt; ++v)
                        map.put(v, rnd.nextInt(valueCnt));

                    System.err.println("Filled");

                    try {
                        Thread.sleep(rnd.nextInt(5000));
                    }
                    catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }, "putter_"+i).start();
        }

        for(int i=0; i<30; ++i) {
            new Thread(() -> {
                while (true) {

                    try {
                        holder[0] = map.get(rnd.nextInt(valueCnt));

                        Thread.sleep(rnd.nextInt(50));
                    }
                    catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }, "getter_"+i).start();
        }

        for(int i=0; i<5; ++i) {
            new Thread(() -> {
                while (true) {
                    try {
                        Thread.sleep(rnd.nextInt(5000));
                    }
                    catch (InterruptedException e) {
                        e.printStackTrace();
                    }

                    System.err.println("Clearing");

                    map.clear();

                    System.err.println("Cleared");
                }
            }, "clearer_"+i).start();
        }

        Thread.sleep(5*60000);
    }
```